# PWD: frameworks/opt/telephony
From 959dcadfec2f7a44f8d884c28f9fa5321ddf6ee0 Mon Sep 17 00:00:00 2001
From: antoniu200 <33966771+antoniu200@users.noreply.github.com>
Date: Sun, 18 Feb 2024 00:29:33 +0200
Subject: [PATCH] Correctly provision carriers that do not offer provisioning:
 ImsPhoneCallTracker.java

Otherwise, we end up either enabling more features than available, or none.

Flamefire: Use isImsCapabilityAvailable and reuse imsManager
---

diff --git a/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java b/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java
index 992d9aa8ea..e05a7da1f5 100644
--- a/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java
+++ b/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java
@@ -3661,22 +3661,38 @@ public class ImsPhoneCallTracker extends CallTracker implements ImsPullCall {
         return (getImsRegistrationTech() == regTech) && mMmTelCapabilities.isCapable(capability);
     }
 
+    private ImsManager getImsManager() {
+        return ImsManager.getInstance(mPhone.getContext(), mPhone.getPhoneId());
+    }
+
     public boolean isVolteEnabled() {
-        boolean isRadioTechLte = getImsRegistrationTech()
-                == ImsRegistrationImplBase.REGISTRATION_TECH_LTE;
-        return isRadioTechLte && mMmTelCapabilities.isCapable(
-                MmTelFeature.MmTelCapabilities.CAPABILITY_TYPE_VOICE);
+        boolean isVolteEnabled = isImsCapabilityAvailable(MmTelFeature.MmTelCapabilities.CAPABILITY_TYPE_VOICE, ImsRegistrationImplBase.REGISTRATION_TECH_LTE);
+        if (isVolteEnabled) {
+            ImsManager imsManager = getImsManager();
+            if (!imsManager.isVolteProvisionedOnDevice()) // Not provisioned, but registered? Carrier probably needs no provisioning.
+                imsManager.setVolteProvisioned(true);
+        }
+        return isVolteEnabled;
     }
 
     public boolean isVowifiEnabled() {
-        boolean isRadioTechIwlan = getImsRegistrationTech()
-                == ImsRegistrationImplBase.REGISTRATION_TECH_IWLAN;
-        return isRadioTechIwlan && mMmTelCapabilities.isCapable(
-                MmTelFeature.MmTelCapabilities.CAPABILITY_TYPE_VOICE);
+        boolean isVowifiEnabled = isImsCapabilityAvailable(MmTelFeature.MmTelCapabilities.CAPABILITY_TYPE_VOICE, ImsRegistrationImplBase.REGISTRATION_TECH_IWLAN);
+        if (isVowifiEnabled) {
+            ImsManager imsManager = getImsManager();
+            if (!imsManager.isWfcProvisionedOnDevice()) // Not provisioned, but registered? Carrier probably needs no provisioning.
+                imsManager.setWfcProvisioned(true);
+        }
+        return isVowifiEnabled;
     }
 
     public boolean isVideoCallEnabled() {
-        return mMmTelCapabilities.isCapable(MmTelFeature.MmTelCapabilities.CAPABILITY_TYPE_VIDEO);
+        boolean isVideoCallEnabled = mMmTelCapabilities.isCapable(MmTelFeature.MmTelCapabilities.CAPABILITY_TYPE_VIDEO);
+        if (isVideoCallEnabled) {
+            ImsManager imsManager = getImsManager();
+            if (!imsManager.isVtProvisionedOnDevice()) // Not provisioned, but registered? Carrier probably needs no provisioning.
+                imsManager.setVtProvisioned(true);
+        }
+        return isVideoCallEnabled;
     }
 
     @Override
